enhancements/20221028-pci-deviceplugin.md

```diff
@@ -0,0 +1,226 @@
+
+In other words, the device plugin advertises the array of devices that are allocatable using `ListAndWatch`, which returns a slice of devices `[]Device{ID, health}`. The IDs will be the PCI address, since this device plugin is specific to the node (kubelet).
+
+If a `PCIDeviceClaim` is deleted, then it's device plugin is signalled that the device is no longer healthy.
+
+The DevicePlugin's gRPC server handles the `Allocate()` request by the kubelet, and prepares the requested device for the virt-launcher container. This `Allocate()` implementation will be called by kubevirt when the `externalResourceProvider` is set to `true`.
+
+The PCIDevicePlugin proposed in this HEP will have an implementation of the `Allocate` RPC request handler that accepts a request for either a number of `resourceName` instances, or a collection of `pciDeviceClaim` names. The deviceplugin will then prepare the devices on that node for the VM.
+
+#### Story 2 in detail
+Carol creates VMs A and B after enabling PCI passthrough on NICs 3 and 4.
+She goes to the VM config UI and clicks the button to add a PCI Device by node and address.
+She clicks on it and sees inputbox/dropdown fields for the PCIDeviceClaim name and an additonal name field. She adds NICs 3 and 4 by setting `pciDeviceClaim` to `janus-000023002` and `name` to `nic3`. Then she saves VM A and then edits the config for VM B, and then adds the device `janus-000023003` as `nic4`.
+
+These devices will be presented in the YAML as:
+
+```yaml
+pciHostDevices:
+  - pciDeviceClaim: janus-000023002
+    name: nic3
+    externalResourceProvider: true
+  - pciDeviceClaim: janus-000023003
+    name: nic4
+    externalResourceProvider: true
+```
+
+### API & UI changes
+
+Currently, in 1.1.0+, the UI for assigning PCI devices to VMs only allows for attaching a single randomly selected device that matches `vendorId:deviceId`:
+
+![](https://raw.githubusercontent.com/tlehman/harvester/master/enhancements/20220722-pci-passthrough/vm-pci-matrix.png)
+
+Proposed change: allow for particular PCIDeviceClaims (whose names include the node name and the PCI Address) to be assigned directly to a VM. This YAML change would 
+
+Previously the YAML configuration for the VM looked like this:
+```yaml
+pciHostDevices:
+  - pciVendorSelector: 10de:1c02
+    resourceName: nvidia.com/GP106_GEFORCE_GTX_1060_3GB
+    externalResourceProvider: false
+  - pciVendorSelector: 10de:10f1
+    resourceName: nvidia.com/GP106_HIGH_DEFINITION_AUDIO_CONTROLLER
+    externalResourceProvider: false
+```
+
+The `externalResourceProvider` was set to false, meaning that kubevirt would _not_ consult an 
```

Comment:
We probably need to look at the upgrade scenario, when users already using pcidevice-controller addon are upgraded to the new version using the device plugin.

Reply:
@ibrokethecloud @tlehman do we still need this?

Reply:
yeah it is needed. this forces kubevirt to leverage the plugin